### ISSUE TITLE
Range hover support

### ISSUE THREAD

**Issue by fannheyward** [fannheyward/coc-rust-analyzer#836]
> the problem is I did find a better way to make Range hover work with nnoremap K. Do you have any ideas?

You're using [getSelectedRange](https://github.com/fannheyward/coc-rust-analyzer/commit/f6092a7a30b9ecd352ebef93479c07c6064cbc79#diff-25d66d74617fe2e23d7946bd6e3ba95640ab1b9bc8947445d604fc271c7c1f12R70) while [expecting Vim to be in visual mode](https://github.com/fannheyward/coc-rust-analyzer/commit/f6092a7a30b9ecd352ebef93479c07c6064cbc79#diff-25d66d74617fe2e23d7946bd6e3ba95640ab1b9bc8947445d604fc271c7c1f12R68), however `:<C-U>` kind of mappings (e.g. `xmap <Plug>(coc-codeaction-selected)`) exit visual mode before executing the range and therefore that check will not pass.

While you could use `<cmd>` mappings to avoid exiting visual mode, exiting is needed to get the visual mode range with `'<` and `'>` like [here](https://github.com/neoclide/coc.nvim/blob/c805bf0a63c67936cf9475cb47cd078280959071/src/workspace.ts#L648) (as far as I'm aware it's not possible to do that while staying in visual mode, you have to exit it first for those marks' positions to be updated).

So, in short, the solution is to make the mapping work like `xmap <Plug>(coc-codeaction-selected)` where you'll detect `visualmode()` *when* the mapping is triggered, not after. [Here's the doCodeAction code](https://github.com/neoclide/coc.nvim/blob/c805bf0a63c67936cf9475cb47cd078280959071/src/handler/codeActions.ts#L85) for reference and you'll see that it fetches the visual range using the marks' positions as mentioned.

_Originally posted by @resolritter in https://github.com/fannheyward/coc-rust-analyzer/issues/256#issuecomment-894335034_

**Comment by fannheyward** [comment]
@resolritter 

> the problem is I did find a better way

Sorry I made a mistake here, what I said is __didn't find a better way__.

rust-analyzer extends the hover request, accepts a position **or range** and returns document. So the range is passed in coc's `doHover` action.

We have this mapping now:

```vim
" Use K to show documentation in preview window.
nnoremap <silent> K :call <SID>show_documentation()<CR>

function! s:show_documentation()
  if (index(['vim','help'], &filetype) >= 0)
    execute 'h '.expand('<cword>')
  elseif (coc#rpc#ready())
    call CocActionAsync('doHover')
  else
    execute '!' . &keywordprg . " " . expand('<cword>')
  endif
endfunction
```

How do we use Range hover alone with basic hover action ?

**Comment by resolritter** [comment]
First of all, to me a range hover only makes sense either in visual mode (the range is the visual area) or using text objects (the range is the area of the text object). In this post I'll be talking only about the visual mode case.

It would work like `xmap <Plug>(coc-codeaction-selected)` and that's the implementation reference you would follow. If you want to use K: `vnoremap <silent> K :<C-u>call CocActionAsync('hoverRange', visualmode())<CR>` 

`<Plug(coc-codeaction-selected)` is defined here: [link](https://github.com/neoclide/coc.nvim/blob/902080fbf7fecd6fc5ff12654a2874d7e109eb2e/plugin/coc.vim#L541). [`cocActionAsync` will forward the `visualmode()` argument from the mapping](https://github.com/neoclide/coc.nvim/blob/902080fbf7fecd6fc5ff12654a2874d7e109eb2e/plugin/coc.vim#L83) to [request_async](https://github.com/neoclide/coc.nvim/blob/902080fbf7fecd6fc5ff12654a2874d7e109eb2e/plugin/coc.vim#L118) and that is how we get [the `mode` parameter in `doCodeAction`](https://github.com/neoclide/coc.nvim/blob/c805bf0a63c67936cf9475cb47cd078280959071/src/plugin.ts#L229). Note that visual mode is received as a function parameter instead of being detected inside of the Node.js code as you're doing in [here](https://github.com/fannheyward/coc-rust-analyzer/commit/f6092a7a30b9ecd352ebef93479c07c6064cbc79#L68); it should be that way because using `<C-U>` mappings will exit visual mode and update the `'<` and `'>` positions, therefore by the point you are doing that [`if (mode)`](https://github.com/fannheyward/coc-rust-analyzer/commit/f6092a7a30b9ecd352ebef93479c07c6064cbc79#L69) it has already exited visual mode and the condition will fail.

coc.vim [already has a `doHover`](https://github.com/neoclide/coc.nvim/blob/c805bf0a63c67936cf9475cb47cd078280959071/src/plugin.ts#L204) but [the implementation](https://github.com/neoclide/coc.nvim/blob/c805bf0a63c67936cf9475cb47cd078280959071/src/handler/hover.ts#L62) seems different. I'm not familiar enough with the code to tell if you would be able to customize it.

**Comment by fannheyward** [comment]
> coc.vim already has a doHover for range 

This range is used for highlight.

I got your point of `visualmode` problem, calls `visualmode` and`getSelectedRange` in coc-rust-analyzer side won't work as expect. The problem is that we need to add action like `hoverRange` in coc side, then forward the mode parameter and get range in Node.js side to use.

**Comment by resolritter** [comment]
If you can't modify the coc.vim source for some reason, then it's possible to use `<cmd>` mappings (Neovim-only IIRC) which does not exit visual mode and therefore your [`if (mode)`](https://github.com/fannheyward/coc-rust-analyzer/blob/f6092a7a30b9ecd352ebef93479c07c6064cbc79/src/client.ts#L69) check would pass. In that case, after the check you would execute some Vim command from Node.js to exit visual mode (needed to get the `'<` and `'>` positions updated, which is what [getSelectedRange](https://github.com/fannheyward/coc-rust-analyzer/blob/f6092a7a30b9ecd352ebef93479c07c6064cbc79/src/client.ts#L70) retrieves).

**Comment by fannheyward** [comment]
> If you can't modify the coc.vim source for some reason

Because the range hover is not in LSP now, IDK we can add it to coc yes or not. Maybe later https://github.com/microsoft/language-server-protocol/issues/377 ?

> it's possible to use <cmd> mappings (Neovim-only IIRC) which does not exit visual mode

How to use `<cmd>` mappings?

**Comment by resolritter** [comment]
For `<cmd>` mappings you'll simply call the function: `xmap K <cmd>call CocAction('doHover')<CR>`

Instead of calling `visualmode()` you should call `mode()`

```vimscript
  fun! EchoMode()
    echo mode() " mode is 'v'
  endfun
  vnoremap <silent> <buffer> K <cmd>call EchoMode()<CR>
```

It finally clicked for me that the middleware `provideHover` can be called with `'doHover'` (I was not understanding how you were going to trigger it before). If this is still open by the end of this week I can try to implement it and submit a pull request.

**Comment by fannheyward** [comment]
> It finally clicked for me that the middleware provideHover can be called with 'doHover'

It's my fault, I should point this out in the beginning.

**Comment by fannheyward** [comment]
@resolritter I've made this PR, can you give it a review?

Also, if you want to test this, you need to patch coc to make it allow float hover in `v` mode:

```patch
diff --git a/src/handler/hover.ts b/src/handler/hover.ts
index 23fcb770..521af161 100644
--- a/src/handler/hover.ts
+++ b/src/handler/hover.ts
@@ -152,7 +152,7 @@ export default class HoverHandler {
     }
     if (target == 'float') {
       let opts: FloatWinConfig = {
-        modes: ['n'],
+        modes: ['n', 'v'],
         maxWidth: this.config.floatMaxWidth,
         maxHeight: this.config.floatMaxHeight,
         autoHide: this.config.autoHide,
```
